Mathlib/Topology/Algebra/Module/Action.lean

```diff
@@ -0,0 +1,253 @@
+containing this topology) which is the definition of continuity.
+
+We also deduce that the module topology is a functor from the category of `R`-modules
+(`R` a topological ring) to the category of topological `R`-modules, and it is perhaps
+unsurprising that this is an adjoint to the forgetful functor. Indeed, if `A` is an `R`-module
+and `M` is a topological `R`-module, then the previous paragraph shows that
+the linear maps `A → M` are precisely the continuous linear maps
+from (`A` with its module topology) to `M`, so the module topology is a left adjoint
+to the forgetful functor.
+
+This file develops the theory of the module topology. We prove that the module topology on
+`R` as a module over itself is `R`'s original topology, and that the module topology
+is isomorphism-invariant.
+
+## Main theorems
+
+* `TopologicalSemiring.toIsModuleTopology : IsModuleTopology R R`. The module
+    topology on `R` is `R`'s topology.
+* `IsModuleTopology.iso [IsModuleTopology R A] (e : A ≃L[R] B) : IsModuleTopology R B`. If `A` and
+    `B` are `R`-modules with topologies, if `e` is a topological isomorphism between them,
+    and if `A` has the module topology, then `B` has the module topology.
+
+## TODO
+
+Forthcoming PRs from the FLT repo will show that the module topology on a (binary or finite) product
+of modules is the product of the module topologies, and that the module topology on the quotient
+of a module `M` is the quotient topology when `M` is equipped with the module topology.
+
+We will also show the slightly more subtle result that if `M`, `N` and `P` are `R`-modules
+equipped with the module topology and if furthermore `M` is finite as an `R`-module,
+then any bilinear map `M × N → P` is continuous.
+
+As a consequence of this, we deduce that if `R` is a commutative topological ring
+and `A` is an `R`-algebra of finite type as `R`-module, then `A` with its module
+topology becomes a topological ring (i.e., multiplication is continuous).
+
+Other TODOs (not done in the FLT repo):
+
+* The module topology is a functor from the category of `R`-modules
+  to the category of topological `R`-modules, and it's an adjoint to the forgetful functor.
+
+-/
+
+section basics
+
+/-
+This section is just boilerplate, defining the module topology and making
+some infrastructure.
+-/
+variable (R : Type*) [TopologicalSpace R] (A : Type*) [Add A] [SMul R A]
+
+/-- The module topology, for a module `A` over a topological ring `R`. It's the finest topology
+making addition and the `R`-action continuous, or equivalently the finest topology making `A`
+into a topological `R`-module. More precisely it's the Inf of the set of
+topologies with these properties; theorems `continuousSMul` and `continuousAdd` show
+that the module topology also has these properties. -/
+abbrev moduleTopology : TopologicalSpace A :=
+  sInf {t | @ContinuousSMul R A _ _ t ∧ @ContinuousAdd A t _}
```

Comment:
Perhaps it's worth a comment here or in the module docstring about how including `ContinuousNeg` is not necessary; maybe because in practice we derive this from `ContinuousSMul`.

Reply:
I would imagine that in the situation where R did not have a neg but M did, there is a chance that neg on M is not continuous. I strongly believe that the mathematical definition of a topological module over a topological ring as being one where add and smul was continuous will have been made without any consideration whatsoever of the semiring case -- semirings are a mathlibism (nat is special-cased in mathematics and nnreal is just a subset of real without any attempt to build a theory on top of it; as far as I know all this semiring stuff is unique to mathlib).

Reply:
Indeed how about R=nnreal, and M=real with the usual topology plus "every subset of (-infty,0] is open". Because the R-action can only push forward, I think M is a topological module in the sense of the natural generalisation of the Wikipedia definition, but neg is not continuous. Aah how about this: I say that *when R is a ring* ContinuousNeg is automatic.

Reply:
OK I've rewritten a bunch of comments and in particular included the observation that neg is continuous in the ring case.